DECISION-RECORD.md

```diff
@@ -0,0 +1,292 @@
+# Trailblaze (September 2020)
+
+TBA
+
+# Prototyping (July 2020)
+
+Research and decision record from the July 2020 prototyping week.
+
+## /entity/item/Q42 vs /entity/Q42
+
+**Decision: /entity/item/Q42 TBA why...**
+
+---
+
+## Consistent "default edit parameters"
+
+**Decision: New edit endpoints should all equally implement certain “default edit parameters” (e.g. token, bot, baserevid, summary, tags).**
+
+In the past various editing interfaces were introduced without a full parameter set.
+Parameters were often then immediately requested but then not added promptly due to other commitments.
+To avoid this, the basic editing parameter set will always be added to edit endpoints.
+
+---
+
+## Pagination
+
+**Decision: Offset parameters (in the Wordpress style) will be used (page, per_page, offset).**
+
+### How others do it
+
+#### Wordpress
+
+- ?page=: page number of results to return
+- ?per_page=: number of records to return in one request, Int(1-100)
+- ?offset=: arbitrary offset at which to start retrieving posts
+
+#### MediaWiki action API
+
+- ?limit and ?continue
+- Different words for combined quieries ?XXlimit ?XXcontinue
+- The word is formed of an abbreviation of the property + ‘limit’ or ‘continue’, 
+- i.e. prop=duplicatefiles (df) params: dflimit, dfcontinue
+
+### Best practices
+
+#### Offset
+
+Retrieves the 20 rows starting with the 100th row
+
+`GET /items?limit=20&offset=100`
+
+Simplest form of paging. Became popular with apps using SQL DBs (LIMIT and OFFSET are part of the SQL SELECT Syntax.
+
+##### Pros
+
+- Easiest to implement, almost no coding required other than passing parameters directly to SQL query.
+- Stateless on the server.
+- Works regardless of custom sort_by parameters
+
+##### Cons
+
+- Not performant for large offset values. The DB needs to scan and count rows starting with 0, and will skip $offset rows.
+- Not consistent when new items are inserted to the table (i.e. Page drift) 
+especially noticeable when ordering items by newest first
+
+#### Keyset
+
+Retrieves the 20 most recent items
+
+`GET /items?limit=20`
+
+Next page, client finds the minimum created date of 2018-01-20T00:00:00 from previously returned results.
+
+`GET /items?limit=20&created:lte:2018-01-20T00:00:00`
+
+##### Pros
+
+- Works with existing filters without additional backend logic. 
+  - Only additional limit parameters needed
+- Consistent ordering even when newer items are inserted into the table. Works well when sorting by most recent first.
+- Consistent performance even with large offsets.
+
+##### Cons
+
+- Tight coupling of paging mechanism to filters and sorting. Forces API users to add filters even if no filters are intended.
+- Does not work for low cardinality fields such as enum strings.
+- Complicated for API users when using custom sort_by fields as the client needs to adjust the filter based on the field used for sorting.
+
+---
+
+## Filtering rows
+
+**Decision: We will use _fields for filtering.**
```

Comment:
This decision doesn't relate to the research below

Reply:
Yeah, so this is an interesting one as _fields is what we currently use, and this was specified in the slide for the smores.
However i can't see the chain of decisions that got us there.

Reply:
I think for the fields we just decided to adopt it as we saw it as an easy choice, and as for the "row filtering" (which is what the research below is actually about) we never made a decision. In any case we addressed it in scope. I guess this would be another reason to move this away from this doc?

Reply:
So I could:
 - Remove this whole section as we are not actually doing it yet (fell out of scope)
OR
 - Leave the research here and remove the decision instead stating some conclusions that we had, so that we dont loose it for the future, and it is easily linkable for others?

Reply:
So I would say, [bookmark](https://www.howtogeek.com/397357/how-to-bookmark-specific-locations-in-a-google-docs-file/) the research in the prototype google doc, and link to it from the `SCOPE.md` document (where we talk about the various filtering rows options) and remove the research from here. But I would also support option 2 depending on the summary :)

Reply:
I dont think we really want to link to / share that full doc publicly? (maybe we do), but this file in my eyes serves as the public extraction of what was done.
I'll go with the option 2 then!